File: c/src/render.c
#include "render.h"
#include "lorentz.h"
#include "rng.h"
#include <math.h>
#include <stdlib.h>
#include <stdatomic.h>
#include <pthread.h>
#include <unistd.h>

#define TILE_SIZE 32

typedef struct { int x0, y0, x1, y1; } Tile;

// Forward declarations
static Spd trace(const Scene *scene, const Camera *cam, int max_depth,
                 double u, double v, Rng *rng);
static Spd trace_world(const Scene *scene, Vec3 origin, Vec3 dir,
                       int depth, Rng *rng);

typedef struct {
    const RenderConfig *cfg;
    const Scene *scene;
    const Camera *cam;
    Tile *tiles;
    int num_tiles;
    atomic_int *next_tile;  // shared across all threads
    uint8_t *pixels;
    int thread_id;
} WorkerCtx;

static void render_tile(const WorkerCtx *ctx, const Tile *tile, Rng *rng) {
    double inv_w = 1.0 / ctx->cfg->width;
    double inv_h = 1.0 / ctx->cfg->height;
    double inv_s = 1.0 / ctx->cfg->samples_per_px;

    for (int y = tile->y0; y < tile->y1; y++) {
        for (int x = tile->x0; x < tile->x1; x++) {
            Spd acc = spd_zero();
            for (int s = 0; s < ctx->cfg->samples_per_px; s++) {
                double jx = rng_float64(rng);
                double jy = rng_float64(rng);
                double u = (x + jx) * inv_w;
                double v = 1.0 - (y + jy) * inv_h;
                Spd sample = trace(ctx->scene, ctx->cam, ctx->cfg->max_depth, u, v, rng);
                spd_add_inplace(&acc, &sample);
            }
            spd_scale_inplace(&acc, inv_s);

            double cx, cy, cz;
            spd_to_xyz(&acc, &cx, &cy, &cz);
            uint8_t r, g, b;
            xyz_to_srgb(cx, cy, cz, &r, &g, &b);

            int idx = (y * ctx->cfg->width + x) * 4;
            ctx->pixels[idx + 0] = r;
            ctx->pixels[idx + 1] = g;
            ctx->pixels[idx + 2] = b;
            ctx->pixels[idx + 3] = 255;
        }
    }
}

static void *worker_func(void *arg) {
    WorkerCtx *ctx = arg;
    Rng rng = rng_seed((uint64_t)ctx->thread_id * 31337);
    while (1) {
        int idx = atomic_fetch_add(ctx->next_tile, 1);
        if (idx >= ctx->num_tiles) break;
        render_tile(ctx, &ctx->tiles[idx], &rng);
    }
    return NULL;
}

uint8_t *render_frame(const RenderConfig *cfg, const Scene *scene, Camera *cam) {
    camera_init(cam);

    int width = cfg->width, height = cfg->height;
    uint8_t *pixels = calloc(width * height * 4, 1);

    // Build tiles
    int nx = (width + TILE_SIZE - 1) / TILE_SIZE;
    int ny = (height + TILE_SIZE - 1) / TILE_SIZE;
    int num_tiles = nx * ny;
    Tile *tiles = malloc(num_tiles * sizeof(Tile));
    int ti = 0;
    for (int y = 0; y < height; y += TILE_SIZE) {
        for (int x = 0; x < width; x += TILE_SIZE) {
            tiles[ti++] = (Tile){
                x, y,
                x + TILE_SIZE < width ? x + TILE_SIZE : width,
                y + TILE_SIZE < height ? y + TILE_SIZE : height
            };
        }
    }

    // Determine thread count
    int num_threads = sysconf(_SC_NPROCESSORS_ONLN);
    if (num_threads < 1) num_threads = 1;

    atomic_int next_tile_counter = 0;

    pthread_t *threads = malloc(num_threads * sizeof(pthread_t));
    WorkerCtx *ctxs = malloc(num_threads * sizeof(WorkerCtx));
    for (int i = 0; i < num_threads; i++) {
        ctxs[i] = (WorkerCtx){
            .cfg = cfg, .scene = scene, .cam = cam,
            .tiles = tiles, .num_tiles = num_tiles,
            .next_tile = &next_tile_counter,
            .pixels = pixels, .thread_id = i,
        };
        pthread_create(&threads[i], NULL, worker_func, &ctxs[i]);
    }
    for (int i = 0; i < num_threads; i++) {
        pthread_join(threads[i], NULL);
    }

    free(threads);
    free(ctxs);
    free(tiles);
    return pixels;
}

// --- Ray tracing ---

static Spd trace(const Scene *scene, const Camera *cam, int max_depth,
                 double u, double v, Rng *rng) {
    Vec3 dir_obs = camera_ray_dir(cam, u, v);
    AberrationResult ab = lorentz_aberrate(dir_obs, cam->beta);

    Spd spd = trace_world(scene, cam->position, ab.dir, max_depth, rng);
    spd = spd_shift(&spd, 1.0 / ab.doppler);
    spd_scale_inplace(&spd, ab.doppler * ab.doppler * ab.doppler);
    return spd;
}

static Spd trace_world(const Scene *scene, Vec3 origin, Vec3 dir,
                       int depth, Rng *rng) {
    if (depth <= 0) return spd_zero();

    Hit hit;
    const Material *mat;
    if (!scene_intersect(scene, origin, dir, 0.001, 1e12, &hit, &mat)) {
        return sky_eval(&scene->sky, dir);
    }

    Spd emitted = material_emitted(mat, &hit);

    // Direct lighting
    Spd direct = spd_zero();
    for (int i = 0; i < scene->num_lights; i++) {
        Vec3 to_light = vec3_sub(scene->lights[i].position, hit.point);
        double dist = vec3_length(to_light);
        Vec3 light_dir = vec3_scale(to_light, 1.0 / dist);

        Hit shadow_hit;
        const Material *shadow_mat;
        if (scene_intersect(scene, hit.point, light_dir, 0.001, dist - 0.001,
                           &shadow_hit, &shadow_mat))
            continue;

        double cos_theta = vec3_dot(hit.normal, light_dir);
        if (cos_theta <= 0) continue;

        double falloff = cos_theta / (4.0 * M_PI * dist * dist);
        Spd light_contrib = scene->lights[i].emission;
        spd_scale_inplace(&light_contrib, falloff);
        spd_add_inplace(&direct, &light_contrib);
    }

    ScatterResult scatter = material_scatter(mat, dir, &hit, rng);
    spd_mul_inplace(&direct, &scatter.reflectance);

    // Indirect lighting
    if (scatter.scattered && depth > 1) {
        Spd bounced = trace_world(scene, hit.point, scatter.out_dir, depth - 1, rng);
        spd_mul_inplace(&bounced, &scatter.reflectance);
        spd_add_inplace(&direct, &bounced);
    }

    // result = emitted + direct
    spd_add_inplace(&emitted, &direct);

    // Source Doppler for moving objects
    if (vec3_length_sq(hit.source_velocity) > 0) {
        Vec3 n_photon = vec3_normalize(vec3_neg(dir));
        Vec3 beta = hit.source_velocity;
        double gamma = 1.0 / sqrt(1.0 - vec3_length_sq(beta));
        double d_source = 1.0 / (gamma * (1.0 - vec3_dot(beta, n_photon)));
        emitted = spd_shift(&emitted, 1.0 / d_source);
        spd_scale_inplace(&emitted, d_source * d_source * d_source);
    }

    return emitted;
}

File: rust/src/render.rs
use std::f64::consts::PI;

use image::{Rgba, RgbaImage};
use rand::rngs::SmallRng;
use rand::{Rng, SeedableRng};
use rayon::prelude::*;

use crate::camera::Camera;
use crate::lorentz;
use crate::scene::Scene;
use crate::spectrum::{self, Spd};
use crate::vec::Vec3;

pub struct Config {
    pub width: u32,
    pub height: u32,
    pub max_depth: u32,
    pub samples_per_px: u32,
}

const TILE_SIZE: u32 = 32;

struct Tile {
    x0: u32,
    y0: u32,
    x1: u32,
    y1: u32,
    seed: u64,
}

/// Render a single frame using tile-based parallel rendering.
pub fn render_frame(cfg: &Config, scene: &Scene, camera: &Camera) -> RgbaImage {
    let mut img = RgbaImage::new(cfg.width, cfg.height);

    // Build tile list
    let mut tiles = Vec::new();
    let mut tile_idx: u64 = 0;
    let mut y = 0;
    while y < cfg.height {
        let mut x = 0;
        while x < cfg.width {
            tiles.push(Tile {
                x0: x,
                y0: y,
                x1: (x + TILE_SIZE).min(cfg.width),
                y1: (y + TILE_SIZE).min(cfg.height),
                seed: tile_idx * 31337,
            });
            tile_idx += 1;
            x += TILE_SIZE;
        }
        y += TILE_SIZE;
    }

    // Render tiles in parallel, collecting pixel data
    let pixel_data: Vec<(u32, u32, [u8; 4])> = tiles
        .par_iter()
        .flat_map(|tile| {
            let mut rng = SmallRng::seed_from_u64(tile.seed);
            let mut pixels = Vec::new();

            let inv_w = 1.0 / cfg.width as f64;
            let inv_h = 1.0 / cfg.height as f64;
            let inv_s = 1.0 / cfg.samples_per_px as f64;

            for y in tile.y0..tile.y1 {
                for x in tile.x0..tile.x1 {
                    let mut acc = Spd::default();
                    for _ in 0..cfg.samples_per_px {
                        let jx: f64 = rng.random();
                        let jy: f64 = rng.random();
                        let u = (x as f64 + jx) * inv_w;
                        let v = 1.0 - (y as f64 + jy) * inv_h;
                        let spd = trace(scene, camera, cfg.max_depth, u, v, &mut rng);
                        acc = acc.add(spd);
                    }
                    acc = acc.scale(inv_s);

                    let (cx, cy, cz) = acc.to_xyz();
                    let (r, g, b) = spectrum::xyz_to_srgb(cx, cy, cz);
                    pixels.push((x, y, [r, g, b, 255]));
                }
            }
            pixels
        })
        .collect();

    // Write pixels to image
    for (x, y, rgba) in pixel_data {
        img.put_pixel(x, y, Rgba(rgba));
    }

    img
}

/// Trace a single pixel at normalized screen coords (u, v).
fn trace(scene: &Scene, camera: &Camera, max_depth: u32, u: f64, v: f64, rng: &mut SmallRng) -> Spd {
    let dir_obs = camera.ray_dir(u, v);
    let ab = lorentz::aberrate(dir_obs, camera.beta);
    let dir_world = ab.dir;
    let doppler = ab.doppler;

    let mut spd = trace_world(scene, camera.position, dir_world, max_depth, rng);

    spd = spd.shift(1.0 / doppler);
    spd = spd.scale(doppler * doppler * doppler);
    spd
}

/// Recursive world-frame ray tracing.
fn trace_world(scene: &Scene, origin: Vec3, dir: Vec3, depth: u32, rng: &mut SmallRng) -> Spd {
    if depth == 0 {
        return Spd::default();
    }

    let Some((hit, mat)) = scene.intersect(origin, dir, 0.001, 1e12) else {
        return (scene.sky)(dir);
    };

    let emitted = mat.emitted(&hit);

    // Direct lighting from point lights
    let mut direct = Spd::default();
    for light in &scene.lights {
        let to_light = light.position - hit.point;
        let dist = to_light.length();
        let light_dir = to_light * (1.0 / dist);

        // Shadow test
        if scene.intersect(hit.point, light_dir, 0.001, dist - 0.001).is_some() {
            continue;
        }

        let cos_theta = hit.normal.dot(light_dir);
        if cos_theta <= 0.0 {
            continue;
        }

        let falloff = cos_theta / (4.0 * PI * dist * dist);
        direct = direct.add(light.emission.scale(falloff));
    }

    let scatter = mat.scatter(dir, &hit, rng);
    let direct_contrib = direct.mul(scatter.reflectance);

    // Indirect lighting (recursive bounce)
    let mut indirect = Spd::default();
    if scatter.scattered && depth > 1 {
        let bounced = trace_world(scene, hit.point, scatter.out_dir, depth - 1, rng);
        indirect = bounced.mul(scatter.reflectance);
    }

    let mut result = emitted.add(direct_contrib).add(indirect);

    // Source Doppler shift for moving objects
    if hit.source_velocity.length_sq() > 0.0 {
        let beta = hit.source_velocity;
        let n_photon = (-dir).normalize();
        let gamma = 1.0 / (1.0 - beta.length_sq()).sqrt();
        let d_source = 1.0 / (gamma * (1.0 - beta.dot(n_photon)));
        result = result.shift(1.0 / d_source);
        result = result.scale(d_source * d_source * d_source);
    }

    result
}
